packages: adblocker-electron{-preload}: make use of `executeJavaScript`

nullptropy opened the pull request:
Use `WebContents.executeJavaScript` for adblocker script injection.

Preload script injection is subject to CSP, breaking functionality on some websites.
This change uses `executeJavaScript` API to bypass CSP:

- Replace preload injection with IPC to main process
- Add `adblocker-inject-script` handler in main process
- Throw an error in the renderer process if the injected script fails

Fixes #4234

seia-soto commented on packages/adblocker-electron-preload/src/index.ts:
```diff
@@ -56,7 +59,9 @@ if (window === window.top && window.location.href.startsWith('devtools://') ===
     if (scripts) {
       for (const script of scripts) {
-        injectScript(script, document);
+        const result: ScriptInjectionResult = await ipcRenderer.invoke('adblocker-inject-script', script)
```
This is a huge security risk to the library since this code may expose a chance for website to inject arbitrary script: https://www.electronjs.org/docs/latest/tutorial/context-isolation#security-considerations

seia-soto replied:
I think we shouldn't pass any script to the backend. Instead, the whole query and execution process should be done on the backend for the security.

nullptropy replied:
Thank you for catching this. I put together this patch with our security model in mind, where we set `sandbox: true`, `contextIsolation: true`, and `nodeIntegration: false`, so this wasn't a problem. I'll address the general case by moving the script execution to the main world completely.

chrmod reviewed (commented):
can we inject scriplets directly from `get-cosmetic-filters-first` handler?

this will avoid unnecessary communication as what is happening here is:
* 1. preload requests scriptlets by `get-cosmetic-filters-first`
* 2. background matches the scritlets and sends it back to preload
* 3. preload sends scriptlets back to background to perform the injection with `adblocker-inject-script`
* 4. background injects scripts in the main world

We can avoid steps 2 and 3, but injecting the scriptlets directly in the step 1.

nullptropy commented:
@seia-soto @chrmod I tried to streamline the implementation and fix the issues, please take another look at the patch.

seia-soto commented on packages/adblocker-electron-preload/src/index.ts:
```diff
@@ -7,24 +7,20 @@
-if (window === window.top && window.location.href.startsWith('devtools://') === false) {
-  (() => {
-    const enableMutationObserver = ipcRenderer.sendSync('is-mutation-observer-enabled');
+if (window === window.top && !window.location.href.startsWith('devtools://')) {
```
This change is not required, please revert this back to `window === window.top && window.location.href.startsWith('devtools://') === false`. I don't feel the problem but this is just our internal preference to make things more explicit. 👍

seia-soto commented on packages/adblocker-electron/src/index.ts:
```diff
@@ -164,101 +158,65 @@ export class ElectronBlocker extends FiltersEngine {
       hostname,
       url,
 
-      // This needs to be done only once per frame
-      getBaseRules: true,
-      getInjectionRules: true,
-      getExtendedRules: true,
-      getRulesFromHostname: true,
-      getRulesFromDOM: false, // Only done on updates (see `onGetCosmeticFiltersUpdated`)
+      // DOM information, only available for updates
+      classes: msg?.classes,
+      hrefs: msg?.hrefs,
+      ids: msg?.ids,
+
+      // Rules to fetch: true for initial call, false for updates
+      getBaseRules: !msg,
+      getInjectionRules: !msg,
+      getExtendedRules: !msg,
+      getRulesFromHostname: !msg,
+
+      // Only true for update calls when we have DOM information
+      getRulesFromDOM: !!msg,
```
I'd prefer adding a dedicated variable for `!msg`:

```ts
    // `msg` is undefined for the initial call and present for subsequent updates
		const isFirstQuery = msg === undefined
```

I don't know what would be the best variable name for this but the followings can be candidates: `isInitialQuery`, `isFirstRequest`, `isInitialRequest`.

nullptropy replied:
Makes perfect sense, named the variable `isFirstRun`.

seia-soto commented on packages/adblocker-electron/src/index.ts:
```diff
@@ -164,101 +158,65 @@ export class ElectronBlocker extends FiltersEngine {
-      callerContext: {
-        frameId: event.frameId,
-        processId: event.processId,
+      for (const script of scripts) {
+        await event.sender.executeJavaScript(script, true);
+      }
 
-        lifecycle: msg.lifecycle,
-      },
-    });
+      if (extended.length > 0) { }
 
-    if (active === false) {
-      return;
+      return { active: true };
+    } catch (error) {
+      return { active: true, error: String(error) };
     }
```
Here, I wonder we can call promises in parallel instead by using `Promise.allSettled`. In our another repository: `ghostery/ghostery-extension`, we don't wait these: https://github.com/ghostery/ghostery-extension/blob/main/extension-manifest-v3/src/background/adblocker.js#L313

nullptropy replied:
Revised the function behaviour here.

seia-soto commented on packages/adblocker-electron/src/index.ts:
```diff
@@ -164,101 +158,65 @@ export class ElectronBlocker extends FiltersEngine {
       callerContext: {
         frameId: event.frameId,
         processId: event.processId,
+        lifecycle: msg?.lifecycle,
```
If you have a nice use-case to this addition to the caller context, please let us know. It'll help our future direction in the implementation. The caller context declarations you can see in many places were added to help debugging.

nullptropy replied:
Nothing off the top of my head right now, but I'll keep this in mind.

seia-soto commented on packages/adblocker-electron/src/index.ts:
```diff
@@ -164,101 +158,65 @@ export class ElectronBlocker extends FiltersEngine {
 
-  public onIsMutationObserverEnabled = (event: Electron.IpcMainEvent): void => {
-    event.returnValue = this.config.enableMutationObserver;
+  public onIsMutationObserverEnabled = async (_: Electron.IpcMainInvokeEvent): Promise<boolean> => {
+    return this.config.enableMutationObserver;
   };
```
Does this need to be `async`?

nullptropy replied:
The function needs to return a `Promise`, and the `async` keyword wraps the body in a promise automatically here.

seia-soto replied:
Makes sense. Thanks.

seia-soto commented on packages/adblocker-electron-preload/src/index.ts:
```diff
@@ -7,24 +7,20 @@
-  ipcRenderer.send('get-cosmetic-filters', window.location.href, data);
+
+function injectCosmeticFilters(data?: Omit<IBackgroundCallback, 'lifecycle'>): Promise<CosmeticFiltersResponse> {
+  return ipcRenderer.invoke('inject-cosmetic-filters', window.location.href, data);
```
Just a dedicated question besides the core functionality: did you suffer from any problems because of event/handler names without namespacing: e.g. prefixing event names with `adblocker-`. If you think you'll see problems because of event names used in adblocker library, I think prefixing event names is totally fine as it's internal details.

nullptropy replied:
We've didn't run into issues with this, but I can totally see people getting confused because of clashing IPC handlers, I'll prefix the event handlers with `@ghostery/adblocker/` to make this less likely.

seia-soto replied:
I think that's a great idea. Thanks for the collaboration.

seia-soto commented on packages/adblocker-electron-preload/src/index.ts:
```diff
@@ -7,24 +7,20 @@
-    const enableMutationObserver = ipcRenderer.sendSync('is-mutation-observer-enabled');
+if (window === window.top && !window.location.href.startsWith('devtools://')) {
+  (async () => {
+    const enableMutationObserver = await ipcRenderer.invoke('is-mutation-observer-enabled');
```
I won't pick this as a problem but the behavior here slightly changed. @remusao Would it be a problem if we initialise `DOMMonitor` after getting a response from ipc main (the background node script)?

seia-soto replied:
Sorry, I think I confused this with `ACTIVE`. I'm closing this review.

seia-soto commented on packages/adblocker-electron/src/index.ts:
```diff
@@ -58,40 +55,36 @@ export class BlockingContext {
-
-  private readonly onGetCosmeticFiltersFirst: (event: Electron.IpcMainEvent, url: string) => void;
+    msg?: IBackgroundCallback,
+  ) => Promise<{ active: boolean; error?: string }>;
```
We can reuse `CosmeticFiltersResponse` here: `Promise<CosmeticFiltersResponse>`

seia-soto commented on packages/adblocker-electron/src/index.ts:
```diff
@@ -164,101 +158,65 @@ export class ElectronBlocker extends FiltersEngine {
+    event: Electron.IpcMainInvokeEvent,
+    url: string,
+    msg?: IBackgroundCallback,
+  ): Promise<{ active: boolean; error?: string }> => {
```
We can reuse `CosmeticFiltersResponse` here: `Promise<CosmeticFiltersResponse>`

seia-soto commented on packages/adblocker-electron/src/index.ts:
```diff
@@ -164,101 +158,65 @@ export class ElectronBlocker extends FiltersEngine {
-        lifecycle: msg.lifecycle,
-      },
-    });
+      if (extended.length > 0) { }
```
From the original code (before this PR), I found that we don't support extended CSS for electron. Let's disable it from match call for now.

seia-soto commented on packages/adblocker-electron-preload/src/index.ts:
```diff
@@ -68,11 +44,11 @@ if (window === window.top && window.location.href.startsWith('devtools://') ===
-            });
+            const response = await injectCosmeticFilters();
+            if (response.error) throw new Error(`error injecting updated cosmetic filters: ${response.error}`)
+            if (!(ACTIVE = response.active)) unload();
```
Do we want to do unloading here? I don't think we need to unload the dom monitor.

seia-soto replied:
`active` from `FiltersEngine.getCosmeticFilters` means that if we want cosmetic filters or not. This code will not be reached at all.

```ts
    if (this.config.loadCosmeticFilters === false) {
      return {
        active: false,
        extended: [],
        scripts: [],
        styles: '',
      };
    }
```

seia-soto reviewed (changes requested):
Thank you for the contribution. I found some headrooms to improve, but I really appreciate introducing the changes.

nullptropy commented:
@seia-soto Thank you for the review! PTAL.

seia-soto commented on packages/adblocker-electron/src/index.ts:
```diff
@@ -164,101 +166,80 @@ export class ElectronBlocker extends FiltersEngine {
+      // Rules to fetch: true for initial call, false for updates
+      getBaseRules: isFirstRun,
+      getInjectionRules: isFirstRun,
+      getExtendedRules: isFirstRun,
```
Please, set this as `false` not to query extended rules.

seia-soto commented:
I found that we were missing tests from adblocker library from https://github.com/ghostery/adblocker/pull/4302 but this PR doesn't have any relationship with that. Therefore, I think it's fine to continue without additional rebasing.

philipp-classen commented on packages/adblocker-electron-preload/src/index.ts:
```diff
@@ -68,11 +53,11 @@ if (window === window.top && window.location.href.startsWith('devtools://') ===
       'DOMContentLoaded',
       () => {
-        DOM_MONITOR = new DOMMonitor((update) => {
+        DOM_MONITOR = new DOMMonitor(async (update) => {
```
Regarding the error handling: since it is now async, the callback will return a promise. I wonder if this is a problem if the callback failed? From what I see, the caller (DOMMonitor) of the callback does not await the result. It would not see the error, nor wait for the processing here to be completed.

Maybe I'm missing something, but I assume the DOMMonitor is that class here: https://github.com/ghostery/adblocker/blob/40a4c8ba9e738b312284cddc6705b5b2a85864d8/packages/adblocker-content/src/index.ts#L176

chrmod replied:
exactly - DOMMonitor relay on the fact the the it callback will execute sync. It is not only about error handling, but for example for `handleNewFeatures` which from what I see assumes the DOM was affected.

nullptropy replied:
I'll think of a way to address this, thank you for catching it!

chrmod commented on packages/adblocker-electron-preload/src/index.ts:
```diff
@@ -7,24 +7,28 @@
-  (() => {
-    const enableMutationObserver = ipcRenderer.sendSync('is-mutation-observer-enabled');
+  (async () => {
+    const enableMutationObserver = await ipcRenderer.invoke(
```
turning this to async may introduce a noticable delay. 

What is the reason to avoid `ipcRenderer.sendSync`?

nullptropy replied:
This will block the renderer until the call is resolved. Definitely not something we would want in our product. Electron itself does suggest to move away from `sendSync` to `invoke` [0].

[0] https://www.electronjs.org/docs/latest/api/ipc-renderer#ipcrenderersendsyncchannel-args

nullptropy replied:
> turning this to async may introduce a noticable delay.

My understanding of async JavaScript doesn't suggest this. Please correct me if I'm wrong.

chrmod replied:
If this works well in your use-case then we should be good with merging those changes. It is clear that that avoiding blocking of the renderer thread is something beneficial, but we have to be mindful that specifically scriptlet injection is very time sensitive. Some scriptlets have to be injected before the page loads own scripts. Otherwise, some ads wont be blocked.

This is way I strongly suggest to decouple scriplet injection from other cosmetic filter injection as described here https://github.com/ghostery/adblocker/pull/4278#pullrequestreview-2326062789

chrmod commented on packages/adblocker-electron-preload/src/index.ts:
```diff
@@ -7,24 +7,28 @@
 if (window === window.top && window.location.href.startsWith('devtools://') === false) {
-  (() => {
-    const enableMutationObserver = ipcRenderer.sendSync('is-mutation-observer-enabled');
+  (async () => {
```
Not sure if there is a strong reason to make this async. 
For the `injectCosmeticFilters` you can use regular `then`.

nullptropy replied:
I did a general API overhaul to switch to using `invoke` in general. This is a much more modern and overall better approach. Please correct me if I'm wrong.

chrmod commented on packages/adblocker-electron/src/index.ts:
```diff
@@ -164,101 +166,80 @@ export class ElectronBlocker extends FiltersEngine {
+      try {
+        await event.sender.insertCSS(styles, { cssOrigin: 'user' });
+      } catch (error) {
+        return {
```
this is a problem - if you fail to inject CSS, you don't try to inject scriptlets

nullptropy replied:
I'll address this, thank you for catching it.

nullptropy replied:
Seems like we don't need to await on insertCSS, or handle errors, we simply call it and proceed to script injection right now. Please let me know if we should do it in order.

chrmod commented on packages/adblocker-electron/src/index.ts:
```diff
@@ -164,101 +166,80 @@ export class ElectronBlocker extends FiltersEngine {
+      }
+    }
+
+    const scriptResults = await Promise.allSettled(
+      scripts.map((script) => event.sender.executeJavaScript(script, true)),
+    );
```
how about combining all scriptlets into a single script so we can inject all at once?

nullptropy replied:
I'd suggest we create another ticket for this once we merge the changes here. I'd like to keep the scope to fixing #4234.

chrmod replied:
make sense!

chrmod reviewed (changes requested):
Thank you very much for pushing this forward. It is clear that electron deserves improvements. 

This review may take a bit more time as the entire team is catching up with electron changes and we want to be sure we can get most out of your efforts, so please bear with us :)

Was thinking, how about instead of requesting scriptlets from the `preload` script, we push them directly based on [did-navigate](https://www.electronjs.org/docs/latest/api/web-contents#event-did-navigate) and [did-frame-navigate](https://www.electronjs.org/docs/latest/api/web-contents#event-did-frame-navigate) which would make it closer to the way the Ghostery extension does the [scritplet injection](https://github.com/ghostery/ghostery-extension/blob/7da14846035cd3c3e71319f8f25e91c18ea9715e/extension-manifest-v3/src/background/adblocker.js#L422)

With that approach, changes to the other cosmetic filters injection could be minimal.

nullptropy commented:
Thank you for the review @chrmod! I've replied most of the comments, and I think the most important one to address is that `DOMUpdater` expects a sync function.

I’d prefer to handle broader changes in other PRs since the electron package isn't fully functional right now. We rely on `@ghostery/adblocker-electron` in our product, so we'll definitely be here to help maintain the project as issues arise.

chrmod commented on packages/adblocker-electron-preload/src/index.ts:
```diff
@@ -68,11 +53,11 @@ if (window === window.top && window.location.href.startsWith('devtools://') ===
-            getCosmeticsFiltersUpdate({
-              ...update,
-            });
+            const response = await injectCosmeticFilters();
+            if (response.error)
+              throw new Error(`error injecting updated cosmetic filters: ${response.error}`);
```
lets remove the `async` from the function definition, and change this to 

```js
injectCosmeticFilters().then((response) => {
  if (response.error) {
    console.error(`error injecting updated cosmetic filters: ${response.error}`);
  }
});

nullptropy commented:
@chrmod PTAL, updated the branch!

chrmod commented on packages/adblocker-electron-preload/src/index.ts:
```diff
@@ -68,10 +53,12 @@ if (window === window.top && window.location.href.startsWith('devtools://') ===
       'DOMContentLoaded',
       () => {
-        DOM_MONITOR = new DOMMonitor((update) => {
+        DOM_MONITOR = new DOMMonitor(async (update) => {
```
```suggestion
        DOM_MONITOR = new DOMMonitor((update) => {
```

chrmod commented on packages/adblocker-electron-preload/src/index.ts:
```diff
@@ -7,24 +7,28 @@
-  IMessageFromBackground,
-  injectScript,
-} from '@cliqz/adblocker-content';
+type CosmeticFiltersResponse = {
+  active: boolean;
+  error?: string;
+};
```
```suggestion
interface ICosmeticFiltersResponse = {
  active: boolean;
  error?: string;
};
```
To match the naming convention of `IBackgroundCallback`

nullptropy commented:
@chrmod Pushed some changes that hopefully addresses the review. I'm losing the thread a bit, please let me know if the changes conform to what you suggested.

chrmod commented on packages/adblocker-electron-preload/src/index.ts:
```diff
@@ -8,25 +8,22 @@
-  (() => {
-    const enableMutationObserver = ipcRenderer.sendSync('is-mutation-observer-enabled');
+  (async () => {
+    const enableMutationObserver = await ipcRenderer.invoke(
```
I don't think this is needed. It should be fine to always create the DOM_MONITOR but only start it when needed. With this line removed, the wrapping function can be sync again

chrmod commented on packages/adblocker-electron/src/index.ts:
```diff
@@ -164,101 +163,63 @@ export class ElectronBlocker extends FiltersEngine {
 
-    // Inject custom stylesheets
-    this.injectStyles(event.sender, styles);
+    event.sender.insertCSS(styles, { cssOrigin: 'user' });
```
I guess we should keep the previous check and insertCSS only if there are some styles
```
if (styles.length > 0) {
```

chrmod reviewed (changes requested):
I think we need one more change around the async code.

Can you already confirm that this works with youtube.com ads?

nullptropy commented:
@chrmod Committed the suggestions. I can confirm that this works on `youtube.com`.

chrmod commented on packages/adblocker-electron-preload/src/index.ts:
```diff
@@ -70,19 +41,17 @@ if (window === window.top && window.location.href.startsWith('devtools://') ===
-            getCosmeticsFiltersUpdate({
-              ...update,
-            });
+            injectCosmeticFilters();
```
```suggestion
            injectCosmeticFilters({
              ...update,
            });
```

chrmod replied:
```suggestion
            injectCosmeticFilters({
               ...update,
            });
```

chrmod reviewed (changes requested):
Looks like one last thing is missing - I've made a code suggestion

chrmod commented on packages/adblocker-electron-preload/src/index.ts:
```diff
@@ -70,8 +41,8 @@ if (window === window.top && window.location.href.startsWith('devtools://') ===
-            getCosmeticsFiltersUpdate({
-              ...update,
+            injectCosmeticFilters({
+               ...update,
```
```suggestion
              ...update,
```

chrmod replied:
this single whitespace breaks the linter

nullptropy replied:
fixed it, thanks!

chrmod reviewed (changes requested):
This is now merge ready, we only have to fix the linter.

seia-soto commented:
Running CI. 👍

seia-soto commented:
We would like to extend our sincere appreciation and congratulations to you for your valuable contribution to the adblocker library. We are pleased to merge this fix, which will benefit all users, and we are grateful for the time and effort you invested in addressing this issue.

chrmod commented:
Many thanks @nullptropy !

nullptropy commented:
Thanks for being patient with me and for all the hard work you guys put into this library, I really appreciate it! @seia-soto @chrmod

PascalPixel commented:
wonderful work, thank you for your time and effort (ﾉ´ヮ´)ﾉ*: ･ﾟ
